src/infra/src/table/alerts/intermediate.rs

```diff
@@ -0,0 +1,607 @@
+// Copyright 2024 OpenObserve Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Defines intermediate data types for translating between service layer data
+//! structures and the database.
+
+use std::{fmt::Display, str::FromStr};
+
+use config::meta::{
+    alerts::{
+        AggFunction as MetaAggFunction, Aggregation as MetaAggregation,
+        CompareHistoricData as MetaCompareHistoricData, Condition as MetaCondition,
+        FrequencyType as MetaFrequencyType, Operator as MetaOperator, QueryType as MetaQueryType,
+    },
+    search::SearchEventType as MetaSearchEventType,
+    stream::StreamType as MetaStreamType,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+
+use crate::errors::{FromI16Error, FromStrError};
+
+/// Query historic data comparison. Stored in the DB as a JSON object.
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct QueryCompareHistoricData {
+    pub offset: String,
+}
+
+impl From<MetaCompareHistoricData> for QueryCompareHistoricData {
+    fn from(value: MetaCompareHistoricData) -> Self {
+        Self {
+            offset: value.offset,
+        }
+    }
+}
+
+impl From<QueryCompareHistoricData> for MetaCompareHistoricData {
+    fn from(value: QueryCompareHistoricData) -> Self {
+        Self {
+            offset: value.offset,
+        }
+    }
+}
+
+/// Threshold frequency type. Stored in the DB as a 16-bit integere.
+pub enum TriggerFrequencyType {
+    Cron,
+    Seconds,
+}
+
+impl TriggerFrequencyType {
+    const CRON: i16 = 0;
+    const SECONDS: i16 = 1;
+}
+
+impl From<TriggerFrequencyType> for i16 {
+    fn from(value: TriggerFrequencyType) -> Self {
+        match value {
+            TriggerFrequencyType::Cron => TriggerFrequencyType::CRON,
+            TriggerFrequencyType::Seconds => TriggerFrequencyType::SECONDS,
+        }
+    }
+}
+
+impl TryFrom<i16> for TriggerFrequencyType {
+    type Error = FromI16Error;
+
+    fn try_from(value: i16) -> Result<Self, Self::Error> {
+        match value {
+            Self::CRON => Ok(TriggerFrequencyType::Cron),
+            Self::SECONDS => Ok(TriggerFrequencyType::Seconds),
+            _ => Err(FromI16Error {
+                value,
+                ty: "FrequencyType".to_owned(),
+            }),
+        }
+    }
+}
+
```

Comment:
```suggestion
#[repr(i16)]
pub enum TriggerFrequencyType {
    Cron,
    Seconds,
}
```
serde actually has this derive macro called `#[repr(...)]` for this purpose.

Reply:
hmmmm, actually this would require another conversion from string to i16, since json only has value to/from String. maybe not since you already implemented the conversion directly.